java/src/jmri/jmrit/ToolsMenu.java

```diff
@@ -172,6 +191,152 @@ public ToolsMenu() {
         add(new JSeparator());
         // LogixNG menu
         add(new jmri.jmrit.logixng.tools.swing.LogixNGMenu());
+
+        updateProgrammerStatus(null);
+        ConnectionStatus.instance().addPropertyChangeListener((PropertyChangeEvent e) -> {
+            if ((e.getPropertyName().equals("change")) || (e.getPropertyName().equals("add"))) {
+                log.debug("Received property {} with value {} ", e.getPropertyName(), e.getNewValue());
+                updateProgrammerStatus(e);
+            }
+        });
+        InstanceManager.addPropertyChangeListener(InstanceManager.getListPropertyName(AddressedProgrammerManager.class),
+                evt -> {
+                    AddressedProgrammerManager m = (AddressedProgrammerManager) evt.getNewValue();
+                    if (m != null) {
+                        m.addPropertyChangeListener(this::updateProgrammerStatus);
+                    }
+                    updateProgrammerStatus(evt);
+                });
+        InstanceManager.getList(AddressedProgrammerManager.class).forEach(m -> m.addPropertyChangeListener(this::updateProgrammerStatus));
+        InstanceManager.addPropertyChangeListener(InstanceManager.getListPropertyName(GlobalProgrammerManager.class),
+                evt -> {
+                    GlobalProgrammerManager m = (GlobalProgrammerManager) evt.getNewValue();
+                    if (m != null) {
+                        m.addPropertyChangeListener(this::updateProgrammerStatus);
+                    }
+                    updateProgrammerStatus(evt);
+                });
+        InstanceManager.getList(GlobalProgrammerManager.class).forEach(m -> m.addPropertyChangeListener(this::updateProgrammerStatus));
+    }
+
+//    /**
+//     * Enable or disable the service mode programmer menu item
+//     * 
+//     * @param evt  Event carrying the status
+//     */
+//    protected void updateProgrammerStatus(PropertyChangeEvent evt) {
+//        log.debug("Property change {} {}", evt.getPropertyName(), evt.getNewValue());
+//        if (evt.getPropertyName().equals("globalProgrammerAvailable")) {
+//            if (jmri.InstanceManager.getNullableDefault(jmri.GlobalProgrammerManager.class) != null) {
+//                serviceAction.setEnabled((boolean)evt.getNewValue());
+//                log.debug("...Service mode enabled now {}", (boolean)evt.getNewValue());
+//            }
+//        }
+//        if (evt.getPropertyName().equals("addressedModePossible")) {
+//            if (jmri.InstanceManager.getNullableDefault(jmri.AddressedProgrammerManager.class) != null) {
+//                opsAction.setEnabled((boolean)evt.getNewValue());
+//                log.debug("...OPs mode enabled now {}", (boolean)evt.getNewValue());
+//            }
+//        }
+//    }
```

Comment:
Is this old code to be removed or code that might be uncommented and used later?
I have no problem with it being kept, just checking.

Reply:
Oops. Should have been removed.